### ISSUE TITLE
false-positive rust-analyzer diagnostic

### ISSUE THREAD

**Issue by GopherJ** [rust-analyzer/rust-analyzer#9852]
![image](https://user-images.githubusercontent.com/33961674/128960391-bedc9381-e7f9-4bec-8a82-df48c851ef29.png)

**Comment by lnicola** [comment]
The problem is that we're picking the wrong `get` function.

Does that code involve proc-macro generated code? Can you extract a test case that shows the same problem (by moving `frame_system::Config` and `DbWeight` to a new project)?

**Comment by sho275** [comment]
I also encountered the same problem when using [nalgebra](https://docs.rs/nalgebra/0.29.0/nalgebra/index.html).

Here is an example code for reproduction. 
```rust
use nalgebra::DVector;

fn main() {
    let z: DVector<f32> = nalgebra::DVector::zeros(3);
    println!("{}", z);
}
```

Although `cargo check` does not emit any error, `rust-analyzer` reports that the number of argments is wrong.

<img width="580" alt="rust-analyzer-mismatched-arg-count" src="https://user-images.githubusercontent.com/80821315/129429531-06fb714a-034e-4633-8517-dc34aa9c540b.PNG">

<img width="350" alt="cargo-check-zero-vec" src="https://user-images.githubusercontent.com/80821315/129429533-2b5af784-282c-4de9-a66e-b9ad6266e236.PNG">

**Comment by lnicola** [comment]
@sho275 that's https://github.com/rust-analyzer/rust-analyzer/issues/8654.

Every case where the type inference doesn't work can cause these false positive diagnostics, because we pick the wrong function. But the root cause can be different: maybe the trait solving got a wrong answer, maybe we didn't expand a macro correctly, or something else. In this case it's because we don't support const generics.

**Comment by GopherJ** [comment]
> The problem is that we're picking the wrong `get` function.
> 
> Does that code involve proc-macro generated code? Can you extract a test case that shows the same problem (by moving `frame_system::Config` and `DbWeight` to a new project)?

it doesn't involve proc-macro generated code

**Comment by flodiebold** [comment]
Can you link to the `get` function that it's resolving to right now, and the `get` function that you'd expect to be used?

**Comment by Chartas** [comment]
I've got the same false positive error and here is a minimal example on how to reproduce it.
The first new will work correctly, the second new will produce `expected 1 argument, found 2 rust-analyzer (mismatched-arg-count)` as a lint.
```Rust
struct MyStruct<const S: usize> { vec: [usize; S] }

impl MyStruct<1> {
    fn new(a: usize) -> Self {
        MyStruct{ vec: [a] }
    }
}

impl MyStruct<2> {
    fn new(a: usize, b: usize) -> Self {
        MyStruct{ vec: [a, b] }
    }
}

pub fn example_error() {
    let _first = MyStruct::<1>::new(0);
    let _second = MyStruct::<2>::new(0, 0);
}
```

**Comment by lnicola** [comment]
@Chartas const generics are not supported.

**Comment by flodiebold** [comment]
That's just #8655, which is not the same root cause as this one.

**Comment by ChayimFriedman2** [comment]
Probably not relevant by now.
